pytorch_pfn_extras/nn/modules/lazy.py

```diff
@@ -90,24 +90,36 @@ def _lazy_load_hook(
         for the details of the hook specification.
         """
         for name in self.lazy_buffer_names:
-            # Avoid shape mismatch error when loading an initialized buffer
-            # onto an uninitialized module instance.
-            self.register_buffer(name, state_dict[prefix + name])
+            key = prefix + name
+            module_initialized = getattr(self, name).shape != (0,)
+            state_initialized = (
+                 key in state_dict and (state_dict[key].shape != (0,)))
+            if module_initialized != state_initialized:
+                raise RuntimeError(
+                    'Can\'t load initialized buffers in non-initialized '
+                    'modules and vice versa')
+            else:
+                # For compatibility with older snapshots
+                if key not in state_dict and not module_initialized:
+                    buffer = torch.Tensor([])
+                    state_dict[key] = buffer
 
         for name in self.lazy_parameter_names:
             # The parameter may not exist in the loaded ``state_dict`` if the
             # original module was serialized before initializing lazy
             # parameters (see comments of ``state_dict``).
             key = prefix + name
-            if key in state_dict:
-                # The model was serialized after initialization.
-                self.register_parameter(
-                    name, torch.nn.Parameter(state_dict[key]))
+            module_initialized = not isinstance(
+                getattr(self, name), UninitializedParameter)
+            state_initialized = key in state_dict
+            if module_initialized != state_initialized:
```

Comment:
I think the only case we have to take care of is this case:
```suggestion
            if module_initialized and not state_initialized:
```
as registering uninitialized parameters to optimizers will raise a warning.